KEP-6063: Update README to match Alpha implementation (#6258)
kubernetes/enhancements: keps/sig-node/6063-pod-pid-limit/README.md

a node-level configuration change.

### Notes/Constraints/Caveats

cgroupsv2 only: This enhancement requires cgroupsv2. On cgroupsv1 nodes, the kubelet will reject pods that specify `spec.resources.limits.pids` during admission, rather than silently ignoring the field. This prevents a situation where the user expects PID enforcement but the node cannot deliver it.

Lower value precedence: When both kubelet and Pod specify limits, the lower value is enforced. This ensures that node-level protections cannot be bypassed by Pod specifications.

Hierarchical cgroup limits: PID limits are hierarchical in cgroups, so the most restrictive limit in the hierarchy applies.

Resource quota and limits: PID limits should be considered in cluster resource planning, though they do not consume schedulable resources like CPU or memory.

Pod Security Admission (PSA) compatibility: The `spec.resources.limits.pids` field has no PSA impact. Pods specifying PID limits are admitted under all PSA profiles (Privileged, Baseline, and Restricted) without triggering any policy violations. PSA does not inspect resource limits, so no special configuration or exemptions are needed.

> Node Declared Features: This feature integrates with the Node Declared Features framework ([KEP-5328](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/5328-node-declared-features)) to handle version skew and mixed cgroupsv1/v2 clusters. Nodes that support per-pod PID limits declare `PerPodPIDLimit` in `node.status.declaredFeatures`, enabling the scheduler to avoid placing pods with `spec.resources.limits.pids` on incompatible nodes.

Reviewer: The declaration is no longer conditioned on the cgroup version, per https://github.com/kubernetes/kubernetes/pull/139277#discussion_r3619557761 `Discover()` only checks the feature gate (https://github.com/kubernetes/kubernetes/blob/f050421a6693073df402fde0cc9f3568087c2d1e/staging/src/k8s.io/component-helpers/nodedeclaredfeatures/features/perpodpidlimit/feature.go#L46-L48), and cgroupsv1 incompatibility is handled at kubelet admission.
```suggestion
Node Declared Features: This feature integrates with the Node Declared Features framework ([KEP-5328](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/5328-node-declared-features)) to handle version skew. Nodes with the `PerPodPIDLimit` feature gate enabled declare it in `node.status.declaredFeatures`, enabling the scheduler to avoid placing pods with `spec.resources.limits.pids` on nodes without the feature gate. cgroupsv1 incompatibility is handled by kubelet admission.
```

#### Feature Gate
The feature is controlled by a new feature gate: `PerPodPIDLimit`.

This feature depends on the following feature gates:

- `PodLevelResources` (Beta, enabled by default since v1.34) — because `pids`
  is specified under `pod.spec.resources.limits`.
- `PodLevelResourcesFixKubeletQOSClass` — fixes QoS class computation when
  pod-level resources are used.

If either dependency is disabled, `PerPodPIDLimit` cannot be enabled — the
kubelet and API server will reject the configuration at startup.

When the feature gate is disabled:

- The kube-apiserver rejects Pods that specify `pids` under `spec.resources.limits`
- Existing Pods that already contain `pids` are preserved and remain unchanged

> - The kubelet ignores the `pids` limit and continues using the node-level `podPidsLimit` setting instead

Reviewer: The implementation went the other way here. When the gate is disabled the kubelet rejects pods that specify `pids` at admission, mirroring the `PodLevelResources` behavior in the same file: https://github.com/kubernetes/kubernetes/blob/f050421a6693073df402fde0cc9f3568087c2d1e/pkg/kubelet/lifecycle/features_linux.go#L62-L70

The NDF section later in this doc also describes the admission recheck on a gate flip, so this bullet currently contradicts both the code and that section.
```suggestion
- The kubelet rejects pods that specify `pids` during admission, mirroring the `PodLevelResources` behavior; already running pods are unaffected until the kubelet restarts and readmits them
```

`node.status.declaredFeatures` during bootstrap. This enables:

- **Scheduler filtering**: The scheduler infers that a pod with
  `spec.resources.limits.pids` requires `PerPodPIDLimit` and only places it on
  nodes that declare the feature. This prevents pods from being scheduled to
  cgroupsv1 nodes or nodes without the feature gate, avoiding kubelet admission

>   rejections that would put the pod into a non-retriable `Failed` state.

Reviewer: Same as my other comments on the cgroupsv2 condition (https://github.com/kubernetes/kubernetes/pull/139277#discussion_r3619557761): the scheduler no longer keeps pods off cgroupsv1 nodes, only off nodes without the feature gate. The intro sentence of this section just above ("enabled and the node supports cgroupsv2") needs the same edit, it is outside the range I can attach a suggestion to.
```suggestion
- **Scheduler filtering**: The scheduler infers that a pod with
  `spec.resources.limits.pids` requires `PerPodPIDLimit` and only places it on
  nodes that declare the feature. This prevents pods from being scheduled to
  nodes without the feature gate, avoiding kubelet admission rejections that
  would put the pod into a non-retriable `Failed` state.
```

This feature touches API validation, field stripping, and kubelet cgroup enforcement. For Alpha, unit test coverage for the following packages is added:

* `pkg/apis/core/validation` will be updated with validation rules for the `pids` resource (range 1024-16384, container-level rejection, `requests.pids` forbidden).
* `pkg/api/pod` will be updated with field stripping logic when the `PerPodPIDLimit` gate is disabled (`requests.pids` is always stripped).
* `pkg/kubelet/cm` will be updated with `getPodPIDLimit` logic (pod limit lower/higher than node, no pod limit fallback), cgroupsv1 rejection error message, `PIDLimitCapped` event emission when the effective limit is capped, and static pod PID limit enforcement (kubelet applies `getPodPIDLimit` uniformly for both regular and static pods).
* `pkg/apis/core/helper` will be updated to exclude `ResourcePID` from `standardContainerResources`.
* `pkg/kubelet/eviction` will be updated to verify that PID pressure detection, eviction ranking, and pod selection are unaffected by `spec.resources.limits.pids` when the `PerPodPIDLimit` gate is enabled.

> * `k8s.io/component-helpers/nodedeclaredfeatures` will be updated to verify that the kubelet declares `PerPodPIDLimit` in `node.status.declaredFeatures` when the feature gate is enabled and cgroupsv2 is available, and does not declare it when the gate is disabled or on cgroupsv1 nodes.

Reviewer: Same review discussion as above (https://github.com/kubernetes/kubernetes/pull/139277#discussion_r3619557761): since `Discover()` no longer checks the cgroup version, the cgroupsv1 scenarios described here have no corresponding tests, `TestPerPodPIDLimitFeature_Discover` covers gate on and gate off only.
```suggestion
* `k8s.io/component-helpers/nodedeclaredfeatures` will be updated to verify that the kubelet declares `PerPodPIDLimit` in `node.status.declaredFeatures` when the feature gate is enabled, and does not declare it when the gate is disabled.
```

* No pod PID limit falls back to node default.
* Pod with pid limit is admitted under Baseline PSA profile and limit is enforced.
* Pod with pid limit is admitted under Restricted PSA profile and limit is enforced.
* Pod specifying `spec.resources.limits.pids` on a cgroupsv1 node is rejected during admission.
* Static pod with `spec.resources.limits.pids` has PID limit enforced at the pod cgroup level (bypasses apiserver).
* Static pod without `spec.resources.limits.pids` falls back to node-level `podPidsLimit`.
* Node enters `PIDPressure` when available PIDs fall below configured eviction thresholds, regardless of pod-level PID limits.
* Eviction manager continues to evict pods under `PIDPressure`; pod-level PID limits do not alter eviction behavior or pod selection.
* Pod-level PID limits reduce aggregate PID consumption and can prevent a workload from triggering `PIDPressure` compared to the node-level default limit.

> * PID pressure calculation is based on node-level PID availability and is unaffected by the configured value of `spec.resources.limits.pids`.

Reviewer: These four scenarios were removed from the implementation per the review feedback in https://github.com/kubernetes/kubernetes/pull/139277#discussion_r3619670099, the coverage moved to the eviction manager unit tests (`TestPIDPressure` and `TestPIDPressureUnaffectedByPerPodPIDLimits`). Listing them here promises e2e tests that do not exist.
```suggestion
```

> * Node with `PerPodPIDLimit` enabled and cgroupsv2 declares `PerPodPIDLimit` in `node.status.declaredFeatures`.

Reviewer: One more instance of the dropped cgroupsv2 condition.
```suggestion
* Node with `PerPodPIDLimit` enabled declares `PerPodPIDLimit` in `node.status.declaredFeatures`.
```

Declared Features). The following version skew scenarios are possible during
rolling upgrades:

**Apiserver ON, kubelet OFF:** The apiserver accepts `spec.resources.limits.pids`.
The kubelet does not declare `PerPodPIDLimit` in `node.status.declaredFeatures`,
so the scheduler avoids placing the pod on this node. The node-level

> `podPidsLimit` applies as a safe fallback.

Reviewer: With the fail closed admission in the implementation this scenario ends in a rejected pod rather than a pod running with just the node default: https://github.com/kubernetes/kubernetes/blob/f050421a6693073df402fde0cc9f3568087c2d1e/pkg/kubelet/lifecycle/features_linux.go#L62-L70

The rollout advice at the end of this section (the other safe fallback sentence) rests on the same assumption and could be updated to match.
```suggestion
**Apiserver ON, kubelet OFF:** The apiserver accepts `spec.resources.limits.pids`.
The kubelet does not declare `PerPodPIDLimit` in `node.status.declaredFeatures`,
so the scheduler avoids placing the pod on this node. If the pod still reaches
the node, kubelet admission rejects it rather than running it without the limit.
```

Declared Features framework
([KEP-5328](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/5328-node-declared-features))
mitigates version skew automatically. Only nodes that declare `PerPodPIDLimit`
receive pods with `spec.resources.limits.pids`, preventing kubelet admission

> rejections that would put the pod into a non-retriable `Failed` state.

Reviewer: With gate only declaration this holds for the feature gate dimension but not for the cgroup version: a cgroupsv1 node with the gate enabled declares the feature, receives the pod and rejects it at admission. The sentence introducing this paragraph ("mixed cgroupsv1/v2 nodes") is part of the same cleanup, it is just above the range I can attach a suggestion to.
```suggestion
mitigates version skew automatically. Only nodes with the feature gate enabled
declare `PerPodPIDLimit` and receive pods with `spec.resources.limits.pids`.
cgroupsv1 incompatibility on such nodes is surfaced as a kubelet admission rejection.
```

    - Impact of its degraded performance or high-error rates on the feature: N/A — cgroup is a kernel interface, not a service.

- `PodLevelResources` feature gate (Beta, enabled by default since v1.34)
  - Usage description: `pids` is specified under `pod.spec.resources.limits`, which requires `PodLevelResources`.
    - Impact of its outage on the feature: If `PodLevelResources` is disabled, `PerPodPIDLimit` cannot be enabled.
    - Impact of its degraded performance or high-error rates on the feature: N/A — feature gate is a binary on/off.

- `PodLevelResourcesFixKubeletQOSClass` feature gate
  - Usage description: Fixes QoS class computation when pod-level resources are used. Without this gate, a pod with only `limits.pids` at pod level could be misclassified as BestEffort.
    - Impact of its outage on the feature: If `PodLevelResourcesFixKubeletQOSClass` is disabled, `PerPodPIDLimit` cannot be enabled.
    - Impact of its degraded performance or high-error rates on the feature: N/A — feature gate is a binary on/off.

- `NodeDeclaredFeatures` feature gate (GA since v1.36)

>   - Usage description: The kubelet declares `PerPodPIDLimit` in `node.status.declaredFeatures` when the feature gate is enabled and cgroupsv2 is available. The scheduler uses this to filter nodes for pods with `spec.resources.limits.pids`.

Reviewer: The cgroupsv2 condition was dropped from discovery per the review discussion in https://github.com/kubernetes/kubernetes/pull/139277#discussion_r3619557761, `Discover()` now only checks the feature gate: https://github.com/kubernetes/kubernetes/blob/f050421a6693073df402fde0cc9f3568087c2d1e/staging/src/k8s.io/component-helpers/nodedeclaredfeatures/features/perpodpidlimit/feature.go#L46-L48

cgroupsv1 incompatibility is handled at kubelet admission instead.
```suggestion
  - Usage description: The kubelet declares `PerPodPIDLimit` in `node.status.declaredFeatures` when the feature gate is enabled. The scheduler uses this to filter nodes for pods with `spec.resources.limits.pids`. cgroupsv1 incompatibility is handled by kubelet admission rather than by the declaration.
```

## Drawbacks

This feature only works on cgroupsv2 nodes. Pods specifying `spec.resources.limits.pids` will be rejected on cgroupsv1 nodes, which may cause confusion in mixed clusters where some nodes have been upgraded to cgroupsv2 and others have not. Workload authors need to be aware of the underlying cgroup version to avoid unexpected admission failures.

> Additionally, the "lower value wins" design means pods can only restrict PID limits below the node default, not raise them. Users who expect `pids: 8192` to override a node `podPidsLimit` of 4096 may find this counterintuitive, though this is consistent with how Kubernetes treats node-level resource limits as the administrator's ceiling.

Reviewer 2: Node default, or node-imposed maximum?
